src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadingCancellationToken, out INamedTypeSymbol? cancellationTokenType))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                if (!(context.ContainingSymbol is IMethodSymbol containingSymbol))
+                {
+                    return;
+                }
+
+                if (!ShouldDiagnose(
+                    invocation,
+                    containingSymbol,
+                    cancellationTokenType,
+                    out string? cancellationTokenParameterName))
+                {
+                    return;
+                }
+
+                // Only underline the method name, not the whole invocation
+                SyntaxNode? expressionNode = GetMethodNameNode(context.Operation.Syntax);
+                if (expressionNode != null)
+                {
+                    context.ReportDiagnostic(expressionNode.CreateDiagnostic(ForwardCancellationTokenToInvocationsRule, cancellationTokenParameterName, invocation.TargetMethod.Name));
+                }
+            },
+            OperationKind.Invocation);
+        }
+
+        // Determines if an invocation should trigger a diagnostic for this rule or not.
+        private static bool ShouldDiagnose(
+            IInvocationOperation invocation,
+            IMethodSymbol containingSymbol,
+            INamedTypeSymbol cancellationTokenType,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            IMethodSymbol method = invocation.TargetMethod;
+
+            // Verify that the current invocation is not passing a token already
+            if (invocation.Arguments.Any(a => a.Parameter.Type.Equals(cancellationTokenType) && !a.IsImplicit))
+            {
+                return false;
+            }
+
+            // Check if the invocation's method has either an optional implicit ct not being used or a params ct parameter not being used or an overload that takes a ct
+            if (!InvocationMethodTakesAToken(method, invocation.Arguments, cancellationTokenType) &&
+                !MethodHasCancellationTokenOverload(method, cancellationTokenType))
+            {
+                return false;
+            }
+
+            // Check if the ancestor method has a ct that we can pass to the invocation
+            if (!VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(cancellationTokenType, GetContainingSymbol(invocation, containingSymbol), out cancellationTokenParameterName))
+            {
+                return false;
+            }
```

Comment:
I think it be valuable to cycle through all the outer lambda/local function/method to find if any one of of them have a cancellation token parameter. It would be fine to offer a code fix only for the case where the immediate enclosing lambda/local function/method has a cancellation token, but a diagnostic for the other cases seems valuable.

```csharp
void CancellableMethod(CancellationToken ct)
{
    LocationFunction();
    return;

    void LocalFunction()
    {
        // This should be flagged to pass in 'ct'
        AnotherCancellableMethod();
    }
}

void AnotherCancellableMethod(CancellationToken ct = default) { }
```

Reply:
Added.